### PR TITLE
`Time#change` should pass the `zone` property through if it's set

### LINKED ISSUE #42467: Time manipulation inconsistency with ActiveSupport::TimeZone
### Steps to reproduce
```ruby
irb(main):001:0> zone = ActiveSupport::TimeZone['Moscow']
irb(main):002:0> zone
=> #<ActiveSupport::TimeZone:0x0000560d29ccf8d0 @name="Moscow", @utc_offset=nil, @tzinfo=#<TZInfo::DataTimezone: Europe/Moscow>>
irb(main):003:0> t1 = Time.new(2021, 5, 29, 0, 0, 0, '+03:00')
irb(main):004:0> t2 = Time.new(2021, 5, 29, 0, 0, 0, zone)
irb(main):005:0> t1
=> 2021-05-29 00:00:00 +0300
irb(main):006:0> t2
=> 2021-05-29 00:00:00 +0300
irb(main):007:0> t1 + 1.days
=> 2021-05-30 00:00:00 +0300
irb(main):008:0> t2 + 1.days
=> 2021-05-30 00:00:00 +0600
```
### Expected behavior
Both times should save time zone after time manipulation

### Actual behavior
Time object created with time zone of Active Support does not save time zone (it uses machine's one) after time manipulation

### System configuration
**Rails version**: 6.0.3.7

**Ruby version**: ruby 2.7.1p83 (2020-03-31 revision a0c7c23c9c) [x86_64-linux]

### PR DESCRIPTION
**Pull request by ghiculescu**
Fixes https://github.com/rails/rails/issues/42467

### CODE REVIEW

File: activesupport/lib/active_support/core_ext/time/calculations.rb [hunk 1]
```diff
@@ -159,6 +159,8 @@ def change(options)
       ::Time.new(new_year, new_month, new_day, new_hour, new_min, new_sec, new_offset)
     elsif utc?
       ::Time.utc(new_year, new_month, new_day, new_hour, new_min, new_sec)
+    elsif zone && zone.is_a?(ActiveSupport::TimeZone)
+      ::Time.new(new_year, new_month, new_day, new_hour, new_min, new_sec, zone)
```
Comment (ghiculescu) [inline activesupport/lib/active_support/core_ext/time/calculations.rb:163]: the original implementation merged this branch and the next one, but `Time.new` doesn't like zones formatted as some kinds of strings, eg. it was crashing on `::Time.new(new_year, new_month, new_day, new_hour, new_min, new_sec, "CST")`

Comment (pixeltrix) [reply activesupport/lib/active_support/core_ext/time/calculations.rb:163]: `time.zone` isn't necessarily an instance of `ActiveSupport::TimeWithZone` - Ruby allows any object there as long as it [quacks in the correct manner][1]. Perhaps a `zone.respond_to?(:utc_to_local)` would be better?

[1]: https://ruby-doc.org/core-3.0.1/Time.html#class-Time-label-Timezone+argument

Comment (ghiculescu) [reply activesupport/lib/active_support/core_ext/time/calculations.rb:163]: That's the docs I was trying to find! I'll tidy it up, thanks.

File: activesupport/lib/active_support/core_ext/time/calculations.rb [hunk 2]
```diff
@@ -159,6 +159,8 @@ def change(options)
       ::Time.new(new_year, new_month, new_day, new_hour, new_min, new_sec, new_offset)
     elsif utc?
       ::Time.utc(new_year, new_month, new_day, new_hour, new_min, new_sec)
+    elsif zone && zone.respond_to?(:utc_to_local)
```
Comment (DNNX) [inline activesupport/lib/active_support/core_ext/time/calculations.rb:162]:
```suggestion
    elsif zone&.respond_to?(:utc_to_local)
```

**Comment by zzak** [comment]
@ghiculescu Can you add a changelog too plz? :bow:

**Comment by ghiculescu** [comment]
👍  changelog added

**Review by zzak — APPROVED** [review]

**Review by pixeltrix — APPROVED** [review]
I suspect that `Time` instances with zone objects has a few more bugs lurking in Active Support 😬

**Pull request merged**